File: Paralelización/Inestables/quickSort.cpp
/* 
  Paralelizacion de Quick Sort con la librería OpenMP
*/
#include <bits/stdc++.h> 
#include <iostream>
#include <omp.h>

using namespace std; 

/* 
    Function swap: this function changes the position in an array, helping with the sorting
    @param a: the first element to be switched 
    @param b: the second element to be switched
*/
void swap(int* a, int* b) 
{ 
	int t = *a; 
	*a = *b; 
	*b = t; 
} 

/* 
  Function Partition: this function is part of the quick sort, it compares the lower value with all the other elements
  from back to front until it finds a smaller value to swap. 
  @param arr:the array of elements to be sorted
  @param low:the first element of the array 
  @param high:the last element of the array 
  @return: position
*/
int partition (int arr[], int low, int high) 
{ 
	int pivot = arr[high];
	int i = (low - 1); 

	for (int j = low; j <= high - 1; j++) 
	{ 
		
		if (arr[j] < pivot) 
		{ 
			i++; 
			swap(&arr[i], &arr[j]); 
		} 
	} 
	swap(&arr[i + 1], &arr[high]); 
	return (i + 1); 
} 

/* 
  Function QuickSort: The goal is to reorganize the array so it is partitioned into two halves, 
  with everything in each side either being less than or greater than our pivot.
  @param arr:the array of elements to be sorted
  @param low: the first element of the array 
  @param high: the last element of the array 
  @return: nothing
*/
void quickSort(int arr[], int low, int high) 
{ 
	if (low < high) 
	{ 
		/* pi is partitioning index, arr[p] is now 
		at right place */
		int pi = partition(arr, low, high); 

		//Esta sección funciona similar a merge sort, se encarga de manejar dos instancias del algoritmo
        //el primero maneja el arreglo sin haber sido ordenado
        //el segundo maneja el arreglo ordenado
        #pragma omp parallel sections num_threads(2)
        {

            #pragma omp section
            { 
                quickSort(arr, low, pi - 1); 
            }
            #pragma omp section
            {
                quickSort(arr, pi + 1, high); 
            }
        }
	} 
} 

/* 
  Function printArray: function that helps us print the array
  @param arr: the array of elements to be sorted
  @param size: size of the array to print
  @return: nothing
*/
void printArray(int arr[], int size) 
{ 
	int i; 
	for (i = 0; i < size; i++) 
		cout << arr[i] << " "; 
	cout << endl; 
} 

int main() 
{ 
    int no = 100;
    int arr[no];
    int nrand =no*10;

    srand((unsigned)time(0));

	//Generación aleatoria de datos
	for(int i = 0; i < no; i++){
	    arr[i] = (rand()%nrand)+1;
	}

    int n = sizeof(arr) / sizeof(arr[0]); 

    cout << "\nArreglo sin ordenar: \n";
    printArray(arr, n);

	quickSort(arr, 0, n - 1); 
	cout << "\nSorted array: \n"; 
	printArray(arr, n); 
	return 0; 
} 

File: backend/Algoritmos/quickSort.js
/* 
  Algoritmo de quickSort en JavaScript
*/

/* 
  Function QuickSort: Algorithm to sort an array by Quick Sort
  @param arr:the array of elements to be sorted
  @param startIndex: the first element of the array 
  @param endIndex: the last element of the array 
  @return: nothing
*/
function quickSort(array, startIndex, endIndex) {
    // verify that the start and end index have not overlapped
    if (startIndex < endIndex) {
      // calculate the pivotIndex
      let pivotIndex = partition(array, startIndex, endIndex)
      // sort the left sub-array
      quickSort(array, startIndex, pivotIndex)
      // sort the right sub-array
      quickSort(array, pivotIndex + 1, endIndex)
    }
  }
  
  /* 
  Function Partition: this function is part of the quick sort, it compares the @param start with all the other elements
  from back to front until it finds a smaller value to swap. 
  @param arr:the array of elements to be sorted
  @param startIndex:the first element of the array 
  @param endIndex:the last element of the array 
  @return:nothing
*/
  function partition(array, startIndex, endIndex) {
    let pivotIndex = Math.floor((startIndex + endIndex) / 2)
    let pivotValue = array[pivotIndex]
  
    while (true) {
      // start at the FIRST index of the sub-array and increment
      // FORWARD until we find a value that is > pivotValue
      while (array[startIndex] < pivotValue) {
        startIndex++
      }
  
      // start at the LAST index of the sub-array and increment
      // BACKWARD until we find a value that is < pivotValue
      while (array[endIndex] > pivotValue) {
        endIndex--
      }
  
      if (startIndex >= endIndex) return endIndex
  
      // swap values at the startIndex and endIndex
      let temp = array[startIndex]
      array[startIndex] = array[endIndex]
      array[endIndex] = temp
    }
  }
  
  let array = [12, 11, 15, 10, 9, 1, 2, 3, 13, 14, 4, 5, 6, 7, 8]
  quickSort(array, 0, array.length - 1)
  console.log(array)